bottlecap/src/traces/trace_agent.rs

```diff
@@ -0,0 +1,203 @@
+const STATS_ENDPOINT_PATH: &str = "/v0.6/stats";
+const INFO_ENDPOINT_PATH: &str = "/info";
+const TRACER_PAYLOAD_CHANNEL_BUFFER_SIZE: usize = 10;
+const STATS_PAYLOAD_CHANNEL_BUFFER_SIZE: usize = 10;
+pub const MAX_CONTENT_LENGTH: usize = 10 * 1024 * 1024;
+
+pub struct TraceAgent {
+    pub config: Arc<config::Config>,
+    pub trace_processor: Arc<dyn trace_processor::TraceProcessor + Send + Sync>,
+    pub trace_flusher: Arc<dyn trace_flusher::TraceFlusher + Send + Sync>,
+    pub stats_processor: Arc<dyn stats_processor::StatsProcessor + Send + Sync>,
+    pub stats_flusher: Arc<dyn stats_flusher::StatsFlusher + Send + Sync>,
+    pub tags_provider: Arc<provider::Provider>,
+}
+
+#[derive(Clone, Copy)]
+pub enum ApiVersion {
+    V04,
+    V05,
+}
+
+impl TraceAgent {
+    pub async fn start_trace_agent(&self) -> Result<(), Box<dyn std::error::Error>> {
+        let now = Instant::now();
+        info!(
+            "Time taken to fetch Trace Agent metadata: {} ms",
+            now.elapsed().as_millis()
+        );
+
+        // setup a channel to send processed traces to our flusher. tx is passed through each
+        // endpoint_handler to the trace processor, which uses it to send de-serialized
+        // processed trace payloads to our trace flusher.
+        let (trace_tx, trace_rx): (Sender<SendData>, Receiver<SendData>) =
+            mpsc::channel(TRACER_PAYLOAD_CHANNEL_BUFFER_SIZE);
+
+        // start our trace flusher. receives trace payloads and handles buffering + deciding when to
+        // flush to backend.
+        let trace_flusher = self.trace_flusher.clone();
+        trace_flusher.start_trace_flusher(trace_rx).await;
+
+        // channels to send processed stats to our stats flusher.
+        let (stats_tx, stats_rx): (
+            Sender<pb::ClientStatsPayload>,
+            Receiver<pb::ClientStatsPayload>,
+        ) = mpsc::channel(STATS_PAYLOAD_CHANNEL_BUFFER_SIZE);
+
+        // start our stats flusher.
+        let stats_flusher = self.stats_flusher.clone();
+        // let stats_config = self.config.clone();
+        tokio::spawn(async move {
+            let stats_flusher = stats_flusher.clone();
+            stats_flusher.start_stats_flusher(stats_rx).await;
+        });
+
+        // setup our hyper http server, where the endpoint_handler handles incoming requests
+        let trace_processor = self.trace_processor.clone();
+        let stats_processor = self.stats_processor.clone();
+        let endpoint_config = self.config.clone();
+        let tags_provider = self.tags_provider.clone();
+
+        let make_svc = make_service_fn(move |_| {
+            let trace_processor = trace_processor.clone();
+            let trace_tx = trace_tx.clone();
+
+            let stats_processor = stats_processor.clone();
+            let stats_tx = stats_tx.clone();
+
+            let endpoint_config = endpoint_config.clone();
+            let tags_provider = tags_provider.clone();
```

Comment:
Rust Q, sooo many clones, can't we just move them all the way down? 🤔

Reply:
nah, need to clone into make_service_fn and then again into service_fn. This all changed in hyper 1.x, so maybe we can look forward to fixing it then?

Reply:
Sg!